driver-core/src/test/java/com/datastax/driver/core/PreparedStatementMultiNodeTest.java

```diff
@@ -0,0 +1,230 @@
+        // as the statement is reprepared when we receive an 'unprepared' response.
+        // uses the Protocol V5 per-query keyspace strategy.
+        executeAfterNodeBroughtBackUp(false, true);
+    }
+
+    @Test(groups = "long")
+    public void should_be_able_to_execute_statement_on_restarted_node_reprepare_on_up_use_keyspace() {
+        // validate that if a node is restarted we should be able to execute a bound statement against it
+        // as the statement is reprepared when the node comes back up.
+        // uses the "USE keyspace" strategy which is an anti-pattern.
+        executeAfterNodeBroughtBackUp(true, false);
+    }
+
+    @Test(groups = "long")
+    public void should_be_able_to_execute_statement_on_restarted_node_use_keyspace() {
+        // validate that if a node is restarted we should be able to execute a bound statement against it
+        // as the statement is reprepared when we receive an 'unprepared' response.
+        // uses the Protocol V5 per-query keyspace strategy.
+        executeAfterNodeBroughtBackUp(false, false);
+    }
+
+    public void executeAfterNodeBroughtBackUp(boolean reprepareOnUp, boolean setKeyspace) {
+        QueryOptions queryOptions = new QueryOptions().setReprepareOnUp(reprepareOnUp).setPrepareOnAllHosts(false);
+        Cluster.Builder builder = createClusterBuilderNoDebouncing()
+                .withNettyOptions(TestUtils.nonQuietClusterCloseOptions)
+                .addContactPointsWithPorts(getContactPointsWithPorts())
+                .withPort(ccm().getBinaryPort())
+                .withQueryOptions(queryOptions)
+                .withLoadBalancingPolicy(new SortingLoadBalancingPolicy());
+
+        // TODO, remove this when V5 is no longer beta.
+        if (setKeyspace) {
+            builder = builder.allowBetaProtocolVersion();
+        }
+
+        Cluster cluster = builder.build();
+        try {
+            Session session = cluster.connect();
+
+            PreparedStatement prep2;
+            PreparedStatement prep3;
+
+            if (setKeyspace) {
+                prep2 = session.prepare(new SimpleStatement("SELECT * from users").setKeyspace(keyspace2));
+                prep3 = session.prepare(new SimpleStatement("SELECT * from users").setKeyspace(keyspace3));
+            } else {
+                // Execute use keyspace before preparing.  This is an anti-pattern but works nonetheless.
+                session.execute("USE " + keyspace2);
+                prep2 = session.prepare(new SimpleStatement("SELECT * from users"));
+                session.execute("USE " + keyspace3);
+                prep3 = session.prepare(new SimpleStatement("SELECT * from users"));
+            }
+
+            Host host1 = TestUtils.findHost(cluster, 1);
+            Host host2 = TestUtils.findHost(cluster, 2);
+
+            // Execute queries 10 times, should always hit host1 due to sorting policy.
+            for (int i = 0; i < 10; i++) {
```

Comment:
Why 10 times?

Reply:
arbitrary, just wanted to hit all hosts.  Will reduce.

Reply:
But doesn't the sorting policy guarantee that you'll always hit host1?

Reply:
ah that's right.  I'm probably using the policy incorrectly then (or not, i need to revisit the tests). I'll look into adjusting one way or another.